### PR TITLE
fix: retry uploads only conditionally

### PR DESCRIPTION
**Pull request by andrewsg**

### CODE REVIEW

File: google/cloud/storage/blob.py [hunk 1]
```diff
@@ -2058,6 +2060,15 @@ def _do_upload(
                   **only** response in the multipart case and it will be the
                   **final** response in the resumable case.
         """
+        if if_metageneration_match is None and num_retries is None:
```
Comment (tritone) [inline google/cloud/storage/blob.py:2063]: I think we should update the docstring as well to make it clear that setting num_retries can be used to override the idempotency requirement (assuming that's the intention), to make sure that users know the implications.

File: tests/unit/test_blob.py [hunk 2]
```diff
@@ -2577,6 +2577,12 @@ def _do_upload_helper(
             if_metageneration_not_match,
             **timeout_kwarg
         )
+
+        # Adjust num_retries expectations to reflect the conditional default in
+        # _do_upload()
+        if num_retries is None and if_metageneration_match is None:
```
Comment (tritone) [inline tests/unit/test_blob.py:2583]: Can you comment on to what extent the conditional is covered by test cases?

Comment (tseaver) [reply tests/unit/test_blob.py:2583]: @tritone the `cover` session under `nox ` requires (since the merge of #308) 100% coverage (branch and statements).

Comment (tritone) [reply tests/unit/test_blob.py:2583]: Ah yeah that's right, good point.

Comment (andrewsg) [reply tests/unit/test_blob.py:2583]: Branch coverage of the code I've added is 100% according to the coverage tool. We are exercising scenarios where if_metageneration_match = true and num_retries is not None.

File: google/cloud/storage/blob.py (hunk 1 above)
Comment (andrewsg) [reply google/cloud/storage/blob.py:2063]: Yes, that's the intention. Will do.

**Review by tseaver — APPROVED** [review]

**Pull request merged**